<a href="https://colab.research.google.com/github/EricAshby/EDA-and-Cleaning-Laptops/blob/main/TEDA1030_Mod6_practice_EricAshby_09_06_23.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Laptop Data Analysis and Cleaning
Eric D. Ashby

##Introduction
This analysis and cleaning project works with the `laptop.csv` file which contains information on laptops sold. Specifically, the data set contains information on the laptops' make, model, and specifications.

The goal of this project is to identify data within the data set in need of cleaning and to apply appropriate data cleaning methods accordingly so as to prepare the data set for machine learning algorithms.

##Purpose of Analysis
This project aims to identify and correct/otherwise deal with:
*  Business Logic Errors
*  Missing Values
*  Standardization Errors
*  Outliers
*  Duplicates
*  Any other forms of dirty data in need of addressing

In [ ]:
import pandas as pd
df = pd.read_csv('laptops.csv')

##Overview

As seen below in the first five rows, the data set contains information on the make, model, and specifications of laptops sold. Right off the bat, missing data is evident in several of the columns.  This will need to be addressed as machine learning algorithms cannot work with null values.

In [ ]:
df.head()

Company   TypeName  Inches screen_resolution  has_ips_panel  is_touchscreen  \
0     NaN  Ultrabook    13.3               NaN              1               0   
1   Apple  Ultrabook    13.3               NaN              0               0   
2      HP   Notebook    15.6           Full HD              0               0   
3   Apple  Ultrabook    15.4               NaN              1               0   
4   Apple  Ultrabook    13.3               NaN              1               0   

   resolution_width  resolution_height cpu_brand cpu_name  cpu_speed   Ram  \
0            1600.0               2560     Intel  Core i5        2.3   8.0   
1             900.0               1440     Intel  Core i5        1.8   8.0   
2            1080.0               1920     Intel  Core i5        2.5   8.0   
3            1800.0               2880     Intel  Core i7        2.7  16.0   
4            1600.0               2560     Intel  Core i5        3.1   8.0   

  memory_type  primary_storage gpu_brand                 gpu_name  OpSys  \
0         SSD              128     Intel   Iris Plus Graphics 640    NaN   
1         NaN              128     Intel         HD Graphics 6000  macos   
2         NaN              256     Intel          HD Graphics 620    NaN   
3         NaN              512       AMD           Radeon Pro 455  macos   
4         SSD              256     Intel   Iris Plus Graphics 650  macos   

   Weight   Price  total_pixels  
0     1.4   71379       4096000  
1     1.3   47896       1296000  
2     1.9   30636       2073600  
3     1.8  135195       5184000  
4     1.4   96096       4096000

In the metadata displayed here, we can identify more columns than we found above with missing data. Also worth noticing, the resolution width being stored as a `float64` (floating point number) rather than an `int64` (integer) is curious. This may be a sign of a standardization or business logic error since the number of pixels is discrete (i.e. and integer).

There are are a total of 1322 entries with 20 columns of data.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1322 entries, 0 to 1321
Data columns (total 20 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Company            1189 non-null   object 
 1   TypeName           1322 non-null   object 
 2   Inches             1322 non-null   float64
 3   screen_resolution  930 non-null    object 
 4   has_ips_panel      1322 non-null   int64  
 5   is_touchscreen     1322 non-null   int64  
 6   resolution_width   1216 non-null   float64
 7   resolution_height  1322 non-null   int64  
 8   cpu_brand          1322 non-null   object 
 9   cpu_name           1322 non-null   object 
 10  cpu_speed          1322 non-null   float64
 11  Ram                1243 non-null   float64
 12  memory_type        390 non-null    object 
 13  primary_storage    1322 non-null   int64  
 14  gpu_brand          1322 non-null   object 
 15  gpu_name           1322 non-null   object 
 16  OpSys              1133 

Focusing back on missing values, the code below more clearly indicates which columns are missing values and how many values are missing. We can clearly see, now, that there are 7 columns with missing data. Of particular note, the Memory Type column is missing over 70% of its data. Unless the machine learning analyses for this data involve memory type, we may consider dropping this column entirely.

In [ ]:
#count null values
print('Number of Missing Values:\n')
print(df.isna().sum())

#calculate percentage of values that are null
print('\n\nPercentage of Values Missing:\n')
print(df.isna().mean()*100)  #multiply by 100 to get percentage instead of decimal form


Number of Missing Values:

Company              133
TypeName               0
Inches                 0
screen_resolution    392
has_ips_panel          0
is_touchscreen         0
resolution_width     106
resolution_height      0
cpu_brand              0
cpu_name               0
cpu_speed              0
Ram                   79
memory_type          932
primary_storage        0
gpu_brand              0
gpu_name               0
OpSys                189
Weight                73
Price                  0
total_pixels           0
dtype: int64


Percentage of Values Missing:

Company              10.060514
TypeName              0.000000
Inches                0.000000
screen_resolution    29.652042
has_ips_panel         0.000000
is_touchscreen        0.000000
resolution_width      8.018154
resolution_height     0.000000
cpu_brand             0.000000
cpu_name              0.000000
cpu_speed             0.000000
Ram                   5.975794
memory_type          70.499244
primary_storage       0.

The code below creates a subset of the data with all rows cotaining even a single null value (i.e. missing value) removed. We are left with a significantly diminished data set with only 193 rows—less that 15% of the original. This represents a massive loss of data. Clearly just dropping entries with missing data right off the bat is not a viable solution (as it rarely, if ever, is).

In [ ]:
df_no_null = df.dropna()
df_no_null

Company            TypeName  Inches screen_resolution  has_ips_panel  \
9       Acer           Ultrabook    14.0           Full HD              1   
20    Lenovo              Gaming    15.6           Full HD              1   
22      Dell  2 in 1 Convertible    13.3           Full HD              0   
26      Dell            Notebook    15.6           Full HD              0   
32      dell           Ultrabook    13.3         Quad HD +              0   
...      ...                 ...     ...               ...            ...   
1294    dell           Ultrabook    12.5           Full HD              0   
1300  lenovo            Notebook    15.6           Full HD              0   
1303  lenovo            Notebook    14.0           Full HD              1   
1305      HP            Notebook    17.3           Full HD              0   
1314  Lenovo            Notebook    13.3           Full HD              1   

      is_touchscreen  resolution_width  resolution_height cpu_brand  \
9                  0            1080.0               1920     Intel   
20                 0            1080.0               1920     Intel   
22                 1            1080.0               1920     Intel   
26                 0            1080.0               1920     Intel   
32                 1            1800.0               3200     Intel   
...              ...               ...                ...       ...   
1294               0            1080.0               1920     Intel   
1300               0            1080.0               1920       AMD   
1303               0            1080.0               1920     Intel   
1305               0            1080.0               1920     Intel   
1314               0            1080.0               1920     Intel   

              cpu_name  cpu_speed   Ram memory_type  primary_storage  \
9              Core i5        1.6  20.0         SSD              256   
20             Core i5        2.5  20.0      Hybrid              128   
22             Core i5        1.6  20.0         SSD              256   
26             Core i5        1.6   8.0         SSD              256   
32             Core i7        1.8  16.0         SSD              512   
...                ...        ...   ...         ...              ...   
1294           Core i7        2.8   8.0         SSD              256   
1300  A12-Series 9720P        3.6  12.0         SSD              512   
1303           Core i5        2.6   8.0         SSD              256   
1305           Core i7        2.7   8.0         SSD              256   
1314           Core i5        2.5   8.0         SSD             -256   

     gpu_brand           gpu_name    OpSys  Weight  Price  total_pixels  
9        Intel   UHD Graphics 620  windows     1.6  41026       2073600  
20      Nvidia   GeForce GTX 1050  windows     2.5  53227       2073600  
22       Intel   UHD Graphics 620  windows     1.6  43636       2073600  
26         AMD         Radeon 530  windows     2.2  42624       2073600  
32       Intel   UHD Graphics 620  windows     1.2  99580       5760000  
...        ...                ...      ...     ...    ...           ...  
1294     Intel    HD Graphics 620  windows     1.6  90043       2073600  
1300       AMD         Radeon 530  windows     2.2  50563       2073600  
1303     Intel    HD Graphics 620  windows     1.3  87912       2073600  
1305    Nvidia      GeForce 930MX  windows     2.6  63936       2073600  
1314     Intel    HD Graphics 620  windows     1.1  53227       2073600  

[193 rows x 20 columns]

This code displays the descriptive statistics for the numeric data in the data set. Glancing through it, we may notice some business logic errors in the minimums of the Price and Primary Storage columns. Negative values do not make sense in these contexts so must be erroneous.

We can spot another peculiarity in the maximum of the Inches column. A 35.6 inch laptop seems more than unlikely. Clearly, this value should be investigated as a potential outlier and, in fact, should be considered a mis-input (business logic error) as a laptop this size just doesn't exist. There may be more outliers in the maxima of some of these columns, such as the very high maximum price, but this remains to be determined later in the analysis.

In [ ]:
df.describe()

Inches  has_ips_panel  is_touchscreen  resolution_width  \
count  1322.000000    1322.000000     1322.000000       1216.000000   
mean     15.122542       0.281392        0.146747       1074.299342   
std       1.945084       0.449848        0.353988        286.711732   
min      10.100000       0.000000        0.000000        768.000000   
25%      14.000000       0.000000        0.000000        900.000000   
50%      15.600000       0.000000        0.000000       1080.000000   
75%      15.600000       1.000000        0.000000       1080.000000   
max      35.600000       1.000000        1.000000       2160.000000   

       resolution_height    cpu_speed          Ram  primary_storage  \
count        1322.000000  1322.000000  1243.000000      1322.000000   
mean         1895.639939     2.301664     9.198713       396.865356   
std           494.741031     0.507130     7.031441       419.775480   
min          1366.000000     0.900000     1.000000     -2048.000000   
25%          1600.000000     2.000000     4.000000       256.000000   
50%          1920.000000     2.500000     8.000000       256.000000   
75%          1920.000000     2.700000     8.000000       512.000000   
max          3840.000000     3.600000    80.000000      2048.000000   

            Weight         Price  total_pixels  
count  1249.000000  1.322000e+03  1.322000e+03  
mean      2.078383  5.838318e+04  2.170972e+06  
std       0.796887  5.045083e+04  1.393255e+06  
min       0.000000 -1.427900e+05  1.049088e+06  
25%       1.500000  3.118475e+04  1.440000e+06  
50%       2.000000  5.109600e+04  2.073600e+06  
75%       2.300000  7.863225e+04  2.073600e+06  
max      11.100000  1.016580e+06  8.294400e+06

Though it is not immediately apparent in the data as we have seen so far, poking around in the non-numeric data reveals that there are errors in the Company column of the data set.

This code provides a look at a the Company column. The code displays the number of entries for each company. Notice that there are duplicate companies with capitalization errors. This is a standardization error that needs fixing and will be addressed later in the analysis.

In [ ]:
df['Company'].value_counts()

Dell         192
Lenovo       190
HP           173
Asus         103
hp            83
dell          81
lenovo        81
Acer          64
asus          35
Toshiba       30
MSI           28
msi           25
acer          24
toshiba       15
Apple         15
Samsung        5
samsung        5
apple          4
Xiaomi         4
Vero           4
Razer          4
Chuwi          3
mediacom       3
Google         3
Mediacom       3
Fujitsu        2
LG             2
Microsoft      2
Huawei         2
microsoft      2
razer          2
Name: Company, dtype: int64

Finally, let's run a quick check for obvious duplicates. We may find that we need to run this check again after we do some cleaning, but for now, the followinf code identifies perfect duplicates in the data. The duplicates discovered in this data set will need to be removed in this analysis.

In [ ]:
print("There are", df.duplicated().sum(), "exact duplicate entries in the data set")
df[df.duplicated()]

There are 10 exact duplicate entries in the data set


Company     TypeName  Inches screen_resolution  has_ips_panel  \
1245    Acer     Notebook    15.6               NaN              0   
1254  Lenovo     Notebook    14.0               NaN              0   
1260    Dell     Notebook    15.6               NaN              0   
1265    Asus     Notebook    15.6               NaN              0   
1270    Asus     Notebook    15.6               NaN              0   
1274      HP  Workstation    17.3               NaN              0   
1285      hp     Notebook    15.6           Full HD              0   
1298   razer    Ultrabook    12.5       4k Ultra HD              1   
1307  Xiaomi     Notebook    15.6           Full HD              1   
1318  Lenovo    Ultrabook    15.6           Full HD              0   

      is_touchscreen  resolution_width  resolution_height cpu_brand  \
1245               0             768.0               1366     Intel   
1254               0             768.0               1366     Intel   
1260               0             768.0               1366     Intel   
1265               0             768.0               1366     Intel   
1270               0             768.0               1366     Intel   
1274               0             900.0               1600     Intel   
1285               0            1080.0               1920       AMD   
1298               1            2160.0               3840     Intel   
1307               0            1080.0               1920     Intel   
1318               0            1080.0               1920     Intel   

            cpu_name  cpu_speed   Ram memory_type  primary_storage gpu_brand  \
1245    Celeron Dual        1.6   4.0         NaN              500     Intel   
1254    Celeron Dual        1.6   2.0         NaN               64     Intel   
1260    Celeron Dual        1.6   2.0         NaN              500     Intel   
1265         Core i7        2.5   4.0         HDD              500    Nvidia   
1270    Celeron Dual        1.6   4.0         NaN              500     Intel   
1274         Core i5        2.8   8.0         NaN              500    Nvidia   
1285  A9-Series 9410        2.9   6.0         NaN             1024       AMD   
1298         Core i7        2.5  16.0         NaN              512     Intel   
1307         Core i7        1.8  16.0         NaN              256    Nvidia   
1318         Core i7        2.7   6.0         NaN              256    Nvidia   

              gpu_name    OpSys  Weight  Price  total_pixels  
1245   HD Graphics 400    linux     2.4  15398       1049088  
1254       HD Graphics  windows     1.5  12201       1049088  
1260       HD Graphics  windows     2.2  20193       1049088  
1265      GeForce 920M  windows     2.2  38379       1049088  
1270       HD Graphics  windows     2.2  19660       1049088  
1274      Quadro M1200  windows     3.1  99154       1440000  
1285    Radeon R7 M440  windows     2.0  29303       2073600  
1298   HD Graphics 620  windows     1.3  95851       8294400  
1307     GeForce MX150      NaN     2.0  74589       2073600  
1318      GeForce 920M  windows     1.5  44276       2073600

##Analysis

###Business Logic Errors

As we have seen above, the data set contains some business logic errors in the form of nonsensically negative values. The folowing code creates a filter to identify entries in the data set with similarly negative values in either of the same columns. The code then displays the flagged entries. We find there to be 107 entries with this business logic error.

In [ ]:
#find entries with negative values in either column
negValFilter = (df['primary_storage'] < 0) | (df['Price'] < 0)

#display aforementioned entries
df[negValFilter]

Company   TypeName  Inches screen_resolution  has_ips_panel  \
5       acer   Notebook    15.6               NaN              0   
19      Dell  Ultrabook    13.3           Full HD              1   
21        hp   Notebook    15.6               NaN              0   
23        HP  Ultrabook    15.6           Full HD              0   
35      acer   Notebook    15.6               NaN              0   
...      ...        ...     ...               ...            ...   
1250      HP    Netbook    11.6               NaN              0   
1282    Dell   Notebook    15.6           Full HD              0   
1286    Asus   Notebook    15.6           Full HD              0   
1308    Dell   Notebook    15.6           Full HD              0   
1314  Lenovo   Notebook    13.3           Full HD              1   

      is_touchscreen  resolution_width  resolution_height cpu_brand  \
5                  0             768.0               1366       AMD   
19                 1            1080.0               1920     Intel   
21                 0               NaN               1366       AMD   
23                 0            1080.0               1920     Intel   
35                 0             768.0               1366     Intel   
...              ...               ...                ...       ...   
1250               0             768.0               1366     Ietnl   
1282               0            1080.0               1920     Ientl   
1286               0            1080.0               1920     Intel   
1308               0            1080.0               1920     Intel   
1314               0            1080.0               1920     Intel   

               cpu_name  cpu_speed   Ram memory_type  primary_storage  \
5        A9-Series 9420        3.0  10.0         HDD              500   
19              Core i5        1.6   8.0         NaN             -128   
21    E-Series E2-9000e        1.5   4.0         NaN              500   
23              Core i7        1.8   8.0         SSD             -256   
35              Core i3        2.7   4.0         NaN            -1024   
...                 ...        ...   ...         ...              ...   
1250       Celeron Dual        1.6   2.0         NaN              -32   
1282            Core i5        1.6   8.0         NaN             -256   
1286            Core i5        2.5   4.0         HDD            -1024   
1308            Core i7        2.7   NaN         SSD             -256   
1314            Core i5        2.5   8.0         SSD             -256   

     gpu_brand           gpu_name    OpSys  Weight  Price  total_pixels  
5          AMD          Radeon R5  windows     2.1 -21312       1049088  
19       Intel   UHD Graphics 620  windows     1.2  52161       2073600  
21         AMD          Radeon R2      NaN     1.9 -13746       1049088  
23       Intel    HD Graphics 620  windows     NaN  35112       2073600  
35       Intel    HD Graphics 620    linux     2.1  19554       1049088  
...        ...                ...      ...     ...    ...           ...  
1250     Intel    HD Graphics 400  windows     1.2  11136       1049088  
1282       AMD         Radeon 530  windows     2.2  42624       2073600  
1286    Nvidia        GeForce 920    linux     2.1  27652       2073600  
1308       AMD     Radeon R5 M430  windows     2.2  39694       2073600  
1314     Intel    HD Graphics 620  windows     1.1  53227       2073600  

[107 rows x 20 columns]

Our first step in correcting these entries will be to flip their signs (i.e. change them from negative to positive). The code below does just this.

In [ ]:
#create filters for negative values in each column, individually
negFilter_Pstorage = df['primary_storage'] < 0
negFilter_price = df['Price'] < 0

#change negative values to positive values
df.loc[negFilter_Pstorage,'primary_storage'] *= -1
df.loc[negFilter_price,'Price'] *= -1

Checking to see if our code worked as expected (reusing our code from above), we find that there are no longer any entires with negative values in the Price or Primary Storage columns.

In [ ]:
negValFilter = (df['primary_storage'] < 0) | (df['Price'] < 0)
df[negValFilter]

Empty DataFrame
Columns: [Company, TypeName, Inches, screen_resolution, has_ips_panel, is_touchscreen, resolution_width, resolution_height, cpu_brand, cpu_name, cpu_speed, Ram, memory_type, primary_storage, gpu_brand, gpu_name, OpSys, Weight, Price, total_pixels]
Index: []

Now, this code recalculates the averages for primary storage and price now that the negative values have been fixed. We see that the average primary storage (447.582) has, predictably, recieved a boost from previous (396.865). Similarly, the average price (61,014.01) has increased from its previous value (58,383.18). These are more accurate descriptors, now that the associated business logic errors have been corrected.

In [ ]:
print("Averages")
print('Primary Storage:', df['primary_storage'].mean())
print('Price:', df['Price'].mean())

Averages
Primary Storage: 447.5824508320726
Price: 61014.00983358548


###Missing Values

As noted in the overview, simply removing entries with missing values is not viable. Instead, we can first use other methods to deal with holes in the data.

####Missing Values in Resolution Width

First, we will look at the missing values in the Resolution Width column. We note that, as can bee seen in the overview, the Resolution Height as well as the Total Pixels columns are not missing any values. Therefore, we can use those two columns to impute the missing data in the Width column.

The code below locates then selects the entries with missing values in the Width column, then fills them with the values calculated by dividing the total pixels by the resolution height (retrieved from their respective columns).

In [ ]:
#fill missing entries in resolution_width with imputed values (total_pixels / resolution_height)
df['resolution_width'].fillna(df['total_pixels'] / df['resolution_height'], inplace = True)

We can check to be sure that all missing values have been filled in with this code. We see that there are no longer any missing values in the Resolution Width column.

In [ ]:
print("Number of Missing Values:\n")
df.isna().sum()

Number of Missing Values:



Company              133
TypeName               0
Inches                 0
screen_resolution    392
has_ips_panel          0
is_touchscreen         0
resolution_width       0
resolution_height      0
cpu_brand              0
cpu_name               0
cpu_speed              0
Ram                   79
memory_type          932
primary_storage        0
gpu_brand              0
gpu_name               0
OpSys                189
Weight                73
Price                  0
total_pixels           0
dtype: int64

The code here calculates and reports the new average for the resolution widths of the laptops in the data set.

In [ ]:
print("Average Resolution Width:", df['resolution_width'].mean())

Average Resolution Width: 1071.428139183056


####Missing Values in Memory Type

As mentioned in the overview, the Memory Type column is missing values for over 70% of the entries. While dropping every entry with missing values is iladvised as it results in a massive data loss, dropping the Memory Type column would only result in losing the memory type data for less than 30% of the entries, alone. Also, as this column does not contain information of particular use to future analyses, it may be appropriate to drop the column entirely, thereby rendering all the entries with missing data in this category useful. Lastly, the data in this column is (mostly*) not imputable, nor is it readily available from the provider of the data set. Therefore, dropping the column is the best method available for dealing with these missing values.

    *It may be possible to determine the memory type from research based on the information in the Company and Type Name (read: Model) columns but such exceeds the scope of this project.

The code below drops the Memory Type column from the data set.

In [ ]:
df.drop(columns = ['memory_type'], inplace = True)
df.head()

Company   TypeName  Inches screen_resolution  has_ips_panel  is_touchscreen  \
0     NaN  Ultrabook    13.3               NaN              1               0   
1   Apple  Ultrabook    13.3               NaN              0               0   
2      HP   Notebook    15.6           Full HD              0               0   
3   Apple  Ultrabook    15.4               NaN              1               0   
4   Apple  Ultrabook    13.3               NaN              1               0   

   resolution_width  resolution_height cpu_brand cpu_name  cpu_speed   Ram  \
0            1600.0               2560     Intel  Core i5        2.3   8.0   
1             900.0               1440     Intel  Core i5        1.8   8.0   
2            1080.0               1920     Intel  Core i5        2.5   8.0   
3            1800.0               2880     Intel  Core i7        2.7  16.0   
4            1600.0               2560     Intel  Core i5        3.1   8.0   

   primary_storage gpu_brand                 gpu_name  OpSys  Weight   Price  \
0              128     Intel   Iris Plus Graphics 640    NaN     1.4   71379   
1              128     Intel         HD Graphics 6000  macos     1.3   47896   
2              256     Intel          HD Graphics 620    NaN     1.9   30636   
3              512       AMD           Radeon Pro 455  macos     1.8  135195   
4              256     Intel   Iris Plus Graphics 650  macos     1.4   96096   

   total_pixels  
0       4096000  
1       1296000  
2       2073600  
3       5184000  
4       4096000

####Remaining Missing Values

Now that the missing values in the Resolution Width and Memory Type have been cleaned, we may expect that most of our missing value problems have been cleared up.

Indeed, we find with the following code, that the majority of the missing value issues have been resolved.

In [ ]:
print("Number of Missing Values:\n")
df.isna().sum()

Number of Missing Values:



Company              133
TypeName               0
Inches                 0
screen_resolution    392
has_ips_panel          0
is_touchscreen         0
resolution_width       0
resolution_height      0
cpu_brand              0
cpu_name               0
cpu_speed              0
Ram                   79
primary_storage        0
gpu_brand              0
gpu_name               0
OpSys                189
Weight                73
Price                  0
total_pixels           0
dtype: int64

Let us again check what the data set would look like if we dropped all the remaining entries with missing values.  The code below accomplishes this. Unlike before, we are left with over seven hundred entries. That's over 50% of the data—not great but far better than the less than 15% that we had before.

In [ ]:
df_no_null = df.dropna()
df_no_null

Company            TypeName  Inches screen_resolution  has_ips_panel  \
9       Acer           Ultrabook    14.0           Full HD              1   
19      Dell           Ultrabook    13.3           Full HD              1   
20    Lenovo              Gaming    15.6           Full HD              1   
22      Dell  2 in 1 Convertible    13.3           Full HD              0   
26      Dell            Notebook    15.6           Full HD              0   
...      ...                 ...     ...               ...            ...   
1312  Lenovo  2 in 1 Convertible    13.3         Quad HD +              1   
1314  Lenovo            Notebook    13.3           Full HD              1   
1317    Asus            Notebook    15.6           Full HD              0   
1318  Lenovo           Ultrabook    15.6           Full HD              0   
1321    Dell              Gaming    15.6           Full HD              0   

      is_touchscreen  resolution_width  resolution_height cpu_brand cpu_name  \
9                  0            1080.0               1920     Intel  Core i5   
19                 1            1080.0               1920     Intel  Core i5   
20                 0            1080.0               1920     Intel  Core i5   
22                 1            1080.0               1920     Intel  Core i5   
26                 0            1080.0               1920     Intel  Core i5   
...              ...               ...                ...       ...      ...   
1312               1            1800.0               3200     Intel  Core i7   
1314               0            1080.0               1920     Intel  Core i5   
1317               0            1080.0               1920     Intel  Core i7   
1318               0            1080.0               1920     Intel  Core i7   
1321               0            1080.0               1920     Intel  Core i7   

      cpu_speed   Ram  primary_storage gpu_brand              gpu_name  \
9           1.6  20.0              256     Intel      UHD Graphics 620   
19          1.6   8.0              128     Intel      UHD Graphics 620   
20          2.5  20.0              128    Nvidia      GeForce GTX 1050   
22          1.6  20.0              256     Intel      UHD Graphics 620   
26          1.6   8.0              256       AMD            Radeon 530   
...         ...   ...              ...       ...                   ...   
1312        2.5  16.0              512     Intel       HD Graphics 520   
1314        2.5   8.0              256     Intel       HD Graphics 620   
1317        2.7   8.0              256     Intel       HD Graphics 620   
1318        2.7   6.0              256    Nvidia          GeForce 920M   
1321        2.8  40.0              128    Nvidia   GeForce GTX 1050 Ti   

        OpSys  Weight  Price  total_pixels  
9     windows     1.6  41026       2073600  
19    windows     1.2  52161       2073600  
20    windows     2.5  53227       2073600  
22    windows     1.6  43636       2073600  
26    windows     2.2  42624       2073600  
...       ...     ...    ...           ...  
1312  windows     1.3  79867       5760000  
1314  windows     1.1  53227       2073600  
1317  windows     2.0  65215       2073600  
1318  windows     1.5  44276       2073600  
1321  windows     2.6  71875       2073600  

[717 rows x 19 columns]

As the remaining columns with missing data are useful to future analyses and as the missing data is neither imputable nor readily available from the provider of the data set, it may now be justifyable to remove the offending entries from the data set to ensure compatibility with machine learning algorithms.

However, this analysis will keep the original data set intact (as presently comprised). A copy of the data set with all null values removed will be saved as `'laptops_cleaned_noNull.csv'` at the end on the analysis.

###Standardization Errors

The code below shows the unique entries in the Company column that remain after removing all remaining entries with missing data. This is much the same as in the overview. The extra companies are created from errors in capitalization.

In [ ]:
print('Companies:')
print(df['Company'].unique())

Companies:
[nan 'Apple' 'HP' 'acer' 'Acer' 'hp' 'Dell' 'apple' 'Lenovo' 'Chuwi'
 'Asus' 'dell' 'lenovo' 'asus' 'MSI' 'microsoft' 'Microsoft' 'Toshiba'
 'msi' 'toshiba' 'Huawei' 'Xiaomi' 'Vero' 'Razer' 'mediacom' 'samsung'
 'Google' 'Mediacom' 'Fujitsu' 'LG' 'Samsung' 'razer']


To fix this standardization error, we use the following code which converts all entries into title case. While this does render some company names somewhat awkward (specifically those with acronyms), it is preferable to a nonstandard naming convention for the purposes of data analysis.

Note that we now no longer have duplicate companies in our list.

In [ ]:
df.loc[:, 'Company'] = df.loc[:, 'Company'].str.title()

print('Companies:')
print(df['Company'].unique())

Companies:
[nan 'Apple' 'Hp' 'Acer' 'Dell' 'Lenovo' 'Chuwi' 'Asus' 'Msi' 'Microsoft'
 'Toshiba' 'Huawei' 'Xiaomi' 'Vero' 'Razer' 'Mediacom' 'Samsung' 'Google'
 'Fujitsu' 'Lg']


Let us once again take a look at the numbers for the Company column.  We see that the errors have been fixed and we get accurate counts for each company in the data set as presently comprised.

In [ ]:
df['Company'].value_counts()

Dell         273
Lenovo       271
Hp           256
Asus         138
Acer          88
Msi           53
Toshiba       45
Apple         19
Samsung       10
Razer          6
Mediacom       6
Microsoft      4
Xiaomi         4
Vero           4
Chuwi          3
Google         3
Huawei         2
Fujitsu        2
Lg             2
Name: Company, dtype: int64

###Outliers

In the overview, we found that the maximum price for laptops seemed suspicious.  Indeed, it seems rather high. To determine if we are justified in out suspicions, the following code calculates z-scores for laptop prices and creates an associated column in the data set.

In [ ]:
#calculate average price
avgPrice = df['Price'].mean()

#calculate standard deviation of price
stdPrice = df['Price'].std()

#create z-score column using z = (price - avgPrice) / stdPrice
df['Price Z-Score'] = (df['Price'] - avgPrice) / stdPrice

The code below identifies and displays the prices outside the threshold of z = 3 (i.e. outside 3 standard deviations from the mean), sorted in descending order of z-score.

In [ ]:
#create filter for entries with prices outside 3 standard deviations from the mean
threeSigmaFilter = (df['Price Z-Score'] >= 3) | (df['Price Z-Score'] <= -3)

#select and display aforementioned entries
df.loc[threeSigmaFilter, ['Price', 'Price Z-Score']].sort_values(by = 'Price Z-Score' , ascending = False)

Price  Price Z-Score
673   1016580      20.230998
765    524810       9.819370
189    324955       5.588091
806    292987       4.911273
595    261019       4.234454
727    233846       3.659155
1036   211788       3.192148
1106   210424       3.163270
229    207259       3.096261

These prices seem suspiciously high, particulary the first entry listed above with a z-score of over 20. Rather than removing these points, however, this analysis will simply note them as imputation is impossible with available information and transformation is outside the scope of this project.

###Duplicates


Now we clear up the duplicates in the data set.

First let's check to see if we gained or lost any duplicates since we started the data cleaning process. We accomplish this with the code below. Displayed are the total number of exact duplicates, the first five of which are also displayed.

In [ ]:
#count the number of exact duplicates, display this number
print("There are", df.duplicated().sum(), "exact duplicate entries in the data set")

#display the first five duplicates
df[df.duplicated()].head()

There are 35 exact duplicate entries in the data set


Company  TypeName  Inches screen_resolution  has_ips_panel  \
975   Toshiba  Notebook    13.3           Full HD              0   
1245     Acer  Notebook    15.6               NaN              0   
1247   Lenovo  Notebook    15.6           Full HD              1   
1248       Hp  Notebook    15.6           Full HD              0   
1251     Asus  Notebook    15.6               NaN              0   

      is_touchscreen  resolution_width  resolution_height cpu_brand  \
975                0            1080.0               1920     Intel   
1245               0             768.0               1366     Intel   
1247               0            1080.0               1920     Intel   
1248               0            1080.0               1920       AMD   
1251               0             768.0               1366     Intel   

            cpu_name  cpu_speed  Ram  primary_storage gpu_brand  \
975          Core i5        2.3  4.0              128     Intel   
1245    Celeron Dual        1.6  4.0              500     Intel   
1247         Core i7        2.6  8.0             1024    Nvidia   
1248  A9-Series 9410        2.9  6.0             1024       AMD   
1251         Core i7        2.5  4.0              500    Nvidia   

               gpu_name    OpSys  Weight  Price  total_pixels  Price Z-Score  
975     HD Graphics 520  windows     1.2  63670       2073600       0.056232  
1245    HD Graphics 400    linux     2.4  15398       1049088      -0.965770  
1247   GeForce GTX 960M  windows     2.6  47899       2073600      -0.277668  
1248     Radeon R7 M440  windows     2.0  29303       2073600      -0.671377  
1251       GeForce 920M  windows     2.2  38379       1049088      -0.479223

As evidenced by both, the increased number of duplicates found and the fact that 4 of the 5 duplicates displayed above were not in our first pass in the overview, we have uncovered more duplicates through our data cleaning process. Specifically, imputation of the Width column and standardization corrections to the Company column are the likely causes.

 Since these are exact duplicates, we can drop them with no loss of data.  However, we have yet to consider near exact duplicates with the following code.  In the code, we consider entries with the same company, cpu, gpu, total pixels, and RAM to be duplicates.

In [ ]:
#create filter for near duplicates (including exact duplicates)
nearDuplicateFilter = df.duplicated(subset = ['Company', 'cpu_name' , 'gpu_name' , 'total_pixels' , 'Ram'])

#count the number of near and/or exact duplicates, display this number
print("There are", nearDuplicateFilter.sum(), "duplicate entries in the data set")

#display the first five duplicates
df[nearDuplicateFilter].head()

There are 579 duplicate entries in the data set


Company   TypeName  Inches screen_resolution  has_ips_panel  \
7    Apple  Ultrabook    13.3               NaN              0   
25   Apple  Ultrabook    13.3               NaN              0   
33   Apple  Ultrabook    13.3               NaN              0   
36    Dell   Notebook    17.3           Full HD              1   
37      Hp   Notebook    15.6               NaN              0   

    is_touchscreen  resolution_width  resolution_height cpu_brand cpu_name  \
7                0             900.0               1440     Intel  Core i5   
25               0             900.0               1440     Intel  Core i5   
33               0             900.0               1440     Intel  Core i5   
36               0            1080.0               1920     Intel  Core i5   
37               0             768.0               1366     Intel  Core i5   

    cpu_speed  Ram  primary_storage gpu_brand           gpu_name    OpSys  \
7         1.8  8.0              256     Intel   HD Graphics 6000    macos   
25        1.6  8.0              128     Intel   HD Graphics 6000    macos   
33        1.6  8.0              256     Intel   HD Graphics 6000    macos   
36        1.6  8.0              128       AMD         Radeon 530  windows   
37        2.5  4.0             1024     Intel    HD Graphics 620  windows   

    Weight  Price  total_pixels  Price Z-Score  
7      1.3  61736       1296000       0.015286  
25     1.4  58555       1296000      -0.052062  
33     1.4  53173       1296000      -0.166008  
36     2.8  52161       2073600      -0.187434  
37     1.9  26037       1049088      -0.740524

A deeper analysis into which near duplicate is most apropriate to keep may be called for in another project, but it exceeds the scope of this analysis. So, for the purposes of this analysis, this project will simply drop all duplicates which appear after the first intance.

The following code deletes all exact and near duplicates from the data set. Displayed is the data set with the aforementioned duplicates removed.

In [ ]:
df.drop_duplicates(subset = ['Company', 'cpu_name' , 'gpu_name' , 'total_pixels' , 'Ram'], inplace = True)
df

Company            TypeName  Inches screen_resolution  has_ips_panel  \
0         NaN           Ultrabook    13.3               NaN              1   
1       Apple           Ultrabook    13.3               NaN              0   
2          Hp            Notebook    15.6           Full HD              0   
3       Apple           Ultrabook    15.4               NaN              1   
4       Apple           Ultrabook    13.3               NaN              1   
...       ...                 ...     ...               ...            ...   
1297  Samsung  2 in 1 Convertible    12.3               NaN              1   
1301      NaN           Ultrabook    15.6           Full HD              1   
1315       Hp           Ultrabook    15.6           Full HD              0   
1320      NaN            Notebook    15.6           Full HD              0   
1321     Dell              Gaming    15.6           Full HD              0   

      is_touchscreen  resolution_width  resolution_height cpu_brand  \
0                  0            1600.0               2560     Intel   
1                  0             900.0               1440     Intel   
2                  0            1080.0               1920     Intel   
3                  0            1800.0               2880     Intel   
4                  0            1600.0               2560     Intel   
...              ...               ...                ...       ...   
1297               1            1600.0               2400   Smsnaug   
1301               0            1080.0               1920     Intel   
1315               0            1080.0               1920     Ietnl   
1320               0            1080.0               1920     Itenl   
1321               0            1080.0               1920     Intel   

            cpu_name  cpu_speed   Ram  primary_storage gpu_brand  \
0            Core i5        2.3   8.0              128     Intel   
1            Core i5        1.8   8.0              128     Intel   
2            Core i5        2.5   8.0              256     Intel   
3            Core i7        2.7  16.0              512       AMD   
4            Core i5        3.1   8.0              256     Intel   
...              ...        ...   ...              ...       ...   
1297  Cortex A72&A53        2.0  10.0               32       ARM   
1301         Core i5        1.6   NaN              256     Intel   
1315         Core i7        2.5   NaN              256       AMD   
1320         Core i5        2.6  20.0              256       AMD   
1321         Core i7        2.8  40.0              128    Nvidia   

                     gpu_name    OpSys  Weight   Price  total_pixels  \
0      Iris Plus Graphics 640      NaN     1.4   71379       4096000   
1            HD Graphics 6000    macos     1.3   47896       1296000   
2             HD Graphics 620      NaN     1.9   30636       2073600   
3              Radeon Pro 455    macos     1.8  135195       5184000   
4      Iris Plus Graphics 650    macos     1.4   96096       4096000   
...                       ...      ...     ...     ...           ...   
1297            Mali T860 MP4    other     1.2   35112       3840000   
1301         UHD Graphics 620      NaN     1.9   58403       2073600   
1315          Radeon R7 M365X  windows     1.8   69104       2073600   
1320           Radeon R7 M465      NaN     1.8   71875       2073600   
1321      GeForce GTX 1050 Ti  windows     2.6   71875       2073600   

      Price Z-Score  
0          0.219445  
1         -0.277731  
2         -0.643155  
3          1.570541  
4          0.742747  
...             ...  
1297      -0.548391  
1301      -0.055280  
1315       0.171279  
1320       0.229946  
1321       0.229946  

[743 rows x 20 columns]

###Polishing the Data Set (Other Dirty Data)

In the overview, we noticed an unusually large monitor size of 35.6 inches.  A quick Google search** indicates that the largest laptops available as of 2022 do not typically exceed 18 inches (novelty items aside). This renders values over 18 in the Inches column, suspect.

Therefore, this analysis notes that the entries displayed by the following code to potentially be business logic errors but does not have sufficient eveidence to alter them at the present time.

      **link at the end of the analysis

In [ ]:
df[df['Inches'] > 18]

Company            TypeName  Inches screen_resolution  has_ips_panel  \
171     Msi              Gaming    18.4           Full HD              0   
411     NaN            Notebook    35.6           Full HD              0   
412    Dell              Gaming    27.3               NaN              1   
426  Lenovo           Ultrabook    24.0               NaN              0   
446     NaN           Ultrabook    33.5               NaN              0   
511  Lenovo            Notebook    24.0           Full HD              0   
577    Dell  2 in 1 Convertible    31.6               NaN              0   
654  Lenovo            Notebook    35.6               NaN              0   

     is_touchscreen  resolution_width  resolution_height cpu_brand  \
171               0            1080.0               1920     Intel   
411               0            1080.0               1920     Intel   
412               0            1440.0               2560     Ietnl   
426               0            1440.0               2560     Intel   
446               1            1504.0               2256     Ientl   
511               0            1080.0               1920     Intel   
577               1             768.0               1366     Intel   
654               0             768.0               1366       AMD   

             cpu_name  cpu_speed   Ram  primary_storage gpu_brand  \
171           Core i7        2.9  32.0              512    Nvidia   
411           Core i3        2.4   8.0             1024    Nvidia   
412           Core i7        2.9  16.0              256    Nvidia   
426           Core i7        2.7  24.0              512     Intel   
446           Core i7        2.5   8.0              256     Intel   
511           Core i5        2.3   8.0              256     Intel   
577            Core M        1.0   NaN              128     Intel   
654  E-Series E2-9000        2.2  10.0              128       AMD   

                    gpu_name    OpSys  Weight   Price  total_pixels  \
171              GTX 980 SLI  windows     NaN  149131       2073600   
411            GeForce 940MX      NaN     2.3   31808       2073600   
412         GeForce GTX 1070  windows     4.4  149131       3686400   
426          HD Graphics 620  windows     1.3  126913       3686400   
446   Iris Plus Graphics 640      NaN     1.3   95851       3393024   
511          HD Graphics 520  windows     2.0   71395       2073600   
577          HD Graphics 615  windows     1.4   32128       1049088   
654       Radeon R2 Graphics  windows     2.2   18595       1049088   

     Price Z-Score  
171       1.865590  
411      -0.618342  
412       1.865590  
426       1.395197  
446       0.737560  
511       0.219784  
577      -0.611567  
654      -0.898084

Also noted in the overview, the data type for the Width column in `float64` instead of the expected `int64`. To investigate this, the code below creates a new column in a copy of the data set that contains the rounded values in the Width column. The code then compares the two columns to determine if they are equal (and, therefore, integers).

In [ ]:
#create a copy of the data set and add a column for the rounded width
df_copy =df.copy()
df_copy['width_rounded'] = df_copy['resolution_width'].round()

#check if Width column contains non integer values (not data types) and create filter
notAnIntFilter = df_copy['resolution_width'] - df_copy['width_rounded'] != 0

#create singular boolean from filter
notAnInt = notAnIntFilter.sum()

#if there are any non-integers contained in the Width column, declare
if notAnInt :
  print("Non-Integer present")
else :
  print("All entries in resolution_width column are integers")

All entries in resolution_width column are integers


We can conclude, then, that there are no business logic errors as expected in this column. Nevertheless, the following code casts the Width column to the expected data type.

In [ ]:
df['resolution_width'] = df['resolution_width'].astype('Int64')

The code below shows us that the type casting was successful.

In [ ]:
df['resolution_width'].info()

<class 'pandas.core.series.Series'>
Int64Index: 743 entries, 0 to 1321
Series name: resolution_width
Non-Null Count  Dtype
--------------  -----
743 non-null    Int64
dtypes: Int64(1)
memory usage: 12.3 KB


##Results

The provided data set has been cleaned of duplicates, outliers, standardization errors, missing values, and business logic errors. Considering outliers and business logic errors, a select few entries are noted in this analysis but not removed.  Additionally, this analysis returns two versions of the cleaned data set: one with all remaining entries with missing values removed and one with those entries left intact.  These are saved as `'laptops_cleaned_noNull.csv'` and `'laptops_cleaned_withNulls.csv'` respectively in the code below.

In [ ]:
df_no_null = df.dropna()
df_no_null.to_csv('laptops_cleaned_noNull.csv')
df.to_csv('laptops_cleaned_withNulls')

##Links

###"What is the maximum laptop size?"

  [https://www.google.com/search?q=what+is+the+maximum+laptop+size&sca_esv=563563987&rlz=1C1UEAD_enUS1027US1027&sxsrf=AB5stBi8CXbKkd8ADmTKbqeJ7a_KRh0HdA%3A1694130782201&ei=XmL6ZP_rC-afur8PyoSJIA&ved=0ahUKEwj_kuf72JmBAxXmj-4BHUpCAgQQ4dUDCBA&uact=5&oq=what+is+the+maximum+laptop+size&gs_lp=Egxnd3Mtd2l6LXNlcnAiH3doYXQgaXMgdGhlIG1heGltdW0gbGFwdG9wIHNpemUyBhAAGBYYHjIGEAAYFhgeMgYQABgWGB4yCBAAGIoFGIYDSJBDULQFWOU0cAF4AZABAJgBngGgAaIQqgEEOC4xMbgBA8gBAPgBAcICChAAGEcY1gQYsAPCAggQABiKBRiRAsICBxAAGIoFGEPCAgUQABiABMICCxAAGIAEGLEDGIMBwgIEEAAYA8ICChAAGIAEGBQYhwLiAwQYACBBiAYBkAYI&sclient=gws-wiz-serp](https://)